### PR TITLE
Initial framework for RPMSG support in i.MX8MP-EVK

### PR DESCRIPTION
**Pull request by emfend**
This PR adds some first steps towards full support of the Cortex M7 found in the i.MX8MP SOC.
These changes make it possible to load the hello world example from the Linux userspace with remoteproc sysfs.
Support for RPMSG via openAMP follows in a separate PR.

### CODE REVIEW

File: samples/hello_world/boards/mimx8mp_evk_itcm.conf [hunk 1]
```diff
@@ -0,0 +1 @@
+CONFIG_OPENAMP_RSC_TABLE=y
```
Comment (arnopo) [inline samples/hello_world/boards/mimx8mp_evk_itcm.conf:1]: Seems that you enable the CONFIG_OPENAMP_RSC_TABLE for bad reason. 
On the stm32mp1 platform I m able to load and run this example without resource table.

No need of the resource table to be able to load the demo, but need should be linked to the redirection of the printk.
On stm32mp1 we use the [RAM CONSOLE](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/st_stm32/stm32mp1/Kconfig.series#L14) that redirects to the Linux remoteproc trace referenced in the resource table. 

In your case if you output the prink on a serial link, you don't need a resource table

**Review by arnopo — CHANGES_REQUESTED** [review]
Hi @emfend

Nice to see the support of a new hardware for this sample.

Few comments but the main point is to clarify the requirement of the resource table.

Could you perform test on your platform to load elf without resource table section? 
There is no reason to not be able to load a firmware without resource table if you use the [remoteproc elf_loader.c](https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/remoteproc_elf_loader.c) routines.
The bypass of the resource table in the linux remoteproc framework is [here](https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/remoteproc_core.c#L1136)

File: lib/open-amp/resource_table.c [hunk 2]
```diff
@@ -33,7 +33,7 @@ extern char ram_console[];
 
 #define __resource Z_GENERIC_SECTION(.resource_table)
 
-#if (CONFIG_OPENAMP_RSC_TABLE_NUM_RPMSG_BUFF > 0) || defined(CONFIG_RAM_CONSOLE)
+#if defined(CONFIG_OPENAMP_RSC_TABLE)
```
[the remark this answers is not in the document]
Comment (emfend) [reply lib/open-amp/resource_table.c:36]: Hello Arnaud,

First of all, thank you for your review.

Even though I haven't tried it, I figured it must work somehow differently on STM32MP1. So I understand the surprise at these changes.
For testing, I'm using a recent (5.10.72) NXP Vendor Linux on the i.MX8MP-EVK. I think this is where most people start and can somehow be taken as a reference.
When I try to load an elf via sysfs without a resource table, the kernel crashes.

[..]
[   83.491674] remoteproc remoteproc0: powering up imx-rproc
[   83.523743] remoteproc remoteproc0: Booting fw image z.elf, size 439236
[   83.530403] remoteproc remoteproc0: No resource table in elf
[   83.536239] Unable to handle kernel NULL pointer dereference at virtual address 0000000000000000
[..]
[   83.721123] Call trace:
[   83.723572]  __memcpy+0x100/0x180
[   83.726889]  rproc_boot+0x340/0x4e0
[   83.730378]  state_store+0x50/0xf0
[   83.733782]  dev_attr_store+0x1c/0x30
[   83.737445]  sysfs_kf_write+0x48/0x60
[   83.741107]  kernfs_fop_write_iter+0x118/0x1ac
[   83.745551]  new_sync_write+0xe8/0x180
[   83.749299]  vfs_write+0x244/0x2a4
[   83.752700]  ksys_write+0x6c/0x100
[   83.756101]  __arm64_sys_write+0x20/0x30
[..]

With an empty resource table, everything works as expected.
I've seen NXP also include an empty resource table in their i.MX8MP Cortex-M7 FreeRTOS samples in MCUXpresso.

[the remark this answers is not in the document]
Comment (arnopo) [reply lib/open-amp/resource_table.c:36]: Look to me a bug in Linux IMX driver...

Here is the the trace I have on stm32mp1 withtthe Hello world sample without resource table.
```
root@stm32mp15-disco-oss:~# echo zephyr.elf >/sys/class/remoteproc/remoteproc0/firmware 
root@stm32mp15-disco-oss:~# echo start >/sys/class/remoteproc/remoteproc0/state 
[ 4852.363587] remoteproc remoteproc0: powering up m4
[ 4852.368458] remoteproc remoteproc0: Booting fw image zephyr.elf, size 504008
[ 4852.374342] remoteproc remoteproc0: no resource table found for this firmware
[ 4852.381550] remoteproc remoteproc0: remote processor m4 is now up
```
It works fine for the [rproc_stm32.c](https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/stm32_rproc.c#L267) driver and the[ imx_rproc.c](https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_rproc.c#L449)  seems having same behavior on lasted Linux kernel.

Would be nice to understand the reason of the crash on Linux first...To be sure that forcing a resource table does not  just mask it.

File: samples/hello_world/boards/mimx8mp_evk_itcm.conf (hunk 1 above)
Comment (emfend) [reply samples/hello_world/boards/mimx8mp_evk_itcm.conf:1]: After spending some time debugging this issue on the Linux side, I can tell that the crash is caused by the vendor-specific imx_rproc driver.
The implementation in the [vendor kernel](https://source.codeaurora.org/external/imx/linux-imx/tree/drivers/remoteproc/imx_rproc.c?h=lf-5.10.y) has some minor modifications to support additional resource table scenarios not present in the mainline kernel.
These changes are responsible for the crash when loading an ELF without a resource table. With a current mainline kernel (5.17-rc2), the firmware can be loaded without any problems.

Unfortunately, there seem to be some other issues when reloading a firmware with a mainline kernel, causing the system to hang when the remote processor boots.
So, at the moment, only the current vendor kernel seems to work for me, but hopefully that will change soon.

TLDR:
- The mainline kernel should not need an RSC table (OPENAMP_RSC_TABLE is not needed).
- The current vendor kernel from NXP needs an RSC table (this looks like a bug to me, but I can't say if it will be fixed or if this is planned for upstream at all).
- Without OPENAMP_RSC_TABLE the hello world example cannot be used with the reference software (NXP imx yocto project).
- Enabling OPENAMP_RSC_TABLE could be seen as a handy helper for people evaluating Zephyr on i.MX with the reference software

What should I do (besides the requested fixes)?
Should I just delete the 'OPENAMP_RSC_TABLE' commit for the hello world example?

Comment (emfend) [reply samples/hello_world/boards/mimx8mp_evk_itcm.conf]: Ok, I decided to just delete the commit that enabled the rsc table for the hello world example. If someone wants to load this example with the current nxp vendor kernel, they have to add the table manually.
With a mainline kernel, the rsc table is not needed.

File: lib/open-amp/resource_table.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (emfend) [reply lib/open-amp/resource_table.c]: Crash caused by a bug in the downstream nxp kernel. Doesn't happen with a newer upstream kernel (also see other comments).

**Review by arnopo — DISMISSED** [review]
Tested with sicess on stm32mp1
Thanks!

File: soc/arm/nxp_imx/mimx8ml8_m7/linker.ld [hunk 3]
```diff
@@ -18,3 +18,13 @@ MEMORY
...
+
+SECTIONS
+     {
+#ifdef CONFIG_OPENAMP_RSC_TABLE
+        SECTION_PROLOGUE(.resource_table,, SUBALIGN(4))
+        {
+            KEEP(*(.resource_table*))
+        } GROUP_LINK_IN(ROMABLE_REGION)
+#endif
+     }
```
Comment (carlocaione) [inline soc/arm/nxp_imx/mimx8ml8_m7/linker.ld:30]: does this work with `CONFIG_CMAKE_LINKER_GENERATOR=y`?

Comment (emfend) [reply soc/arm/nxp_imx/mimx8ml8_m7/linker.ld:30]: Although it is possible to build the program with the generated linker script, the resource table section will not be included then.
I used the same mechanism as in the linker script for stm32mp1 - so there is the same behavior.

So if one decides to use a different linker script, for whatever reason, the table is missing, but one will notice that soon.
To save users from this pitfall, maybe CONFIG_OPENAMP_RSC_TABLE could select CONIG_LD_LINKER_TEMPLATE or depend on !CONFIG_CMAKE_LINKER_GENERATOR.
What do you think?

Comment (tejlmand) [reply soc/arm/nxp_imx/mimx8ml8_m7/linker.ld:30]: > Although it is possible to build the program with the generated linker script, the resource table section will not be included then.

and we should try to keep things aligned, so this is not good.
Today, the `mimx8ml8` does support the CMake linker script generator, try for yourself:
```
cmake -GNinja -DBOARD=mimx8mp_evk_itcm -Bbuild samples/hello_world/ -DCONFIG_CMAKE_LINKER_GENERATOR=y
```

> I used the same mechanism as in the linker script for stm32mp1 - so there is the same behavior.

so that means we now have some work to do:
https://github.com/zephyrproject-rtos/zephyr/issues/43206
but is not a reason for introducing tech-debt in a new PR.

> 
> So if one decides to use a different linker script, for whatever reason, the table is missing, but one will notice that soon. To save users from this pitfall, maybe CONFIG_OPENAMP_RSC_TABLE could select CONIG_LD_LINKER_TEMPLATE or depend on !CONFIG_CMAKE_LINKER_GENERATOR. What do you think?

Linker script generator can also be used for generating ld script 😉 
Is is really that hard to add this line:
```
zephyr_linker_section(NAME resource_table GROUP ROM_REGION NOINPUT)
zephyr_linker_section_configure(SECTION resource_table KEEP INPUT ".resource_table*")
```
here: https://github.com/zephyrproject-rtos/zephyr/blob/9d2d9750b7de4e55c96ace07bb2ae3feabdb3065/soc/arm/nxp_imx/mimx8ml8_m7/CMakeLists.txt

and then test it works ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
please also ensure linker script generator works for ld generated scripts.

File: soc/arm/nxp_imx/mimx8ml8_m7/linker.ld (hunk 3 above)
Comment (emfend) [reply soc/arm/nxp_imx/mimx8ml8_m7/linker.ld:30]: No, it's not difficult indeed. It was just a lack of knowledge regarding the script generator :/

Thanks to your detailed hint, this section is now included in the output in both linker script variants.
A test in a real system was also successful with both methods.

In both cases, the section will not be added if CONFIG_OPENAMP_RSC_TABLE is not selected.

Thanks for your helpful input!

Comment (arnopo) [reply soc/arm/nxp_imx/mimx8ml8_m7/linker.ld:30]: Lack of knoledge also on the script generator...But look strange that we need to duplicate. 

On other side if I suppress one of the declaration either in the linker.ld or the Cmakelists (on stm32mp1 soc), I have `warning: orphan section .resource_table` depending on the CONFIG_CMAKE_LINKER_GENERATOR.

**Review by arnopo — DISMISSED** [review]

**Review by carlocaione — DISMISSED** [review]

File: soc/arm/nxp_imx/mimx8ml8_m7/linker.ld (hunk 3 above)
Comment (tejlmand) [reply soc/arm/nxp_imx/mimx8ml8_m7/linker.ld:30]: > But look strange that we need to duplicate.

this is needed for the time, as the old approach uses the preprocessor to generate the final ld linker script.
The new approach describes everything in CMake, which allows us to then generate linker scripts for different linkers, ld (ld scripts), armlink (scatter files), and more linkers in future.

But it is a transition in progress, and thus needs to be described in both ways for time being.

**Review by tejlmand — DISMISSED** [review]
lgtm

**Comment by emfend** [comment]
Hi @carlocaione,

I received a notification that the footprint delta check failed [1]. It looks like this is caused by a missing device tree header file, but since this looks somehow unrelated to the changes in the PR (at least for me) I don't know how to proceed...
Any advice on this? Thanks!

[1] https://github.com/zephyrproject-rtos/zephyr/runs/5539871424?check_suite_focus=true

**Review by tejlmand — APPROVED** [review]
lgtm - re-approved

**Review by carlocaione — APPROVED** [review]

**Comment by carlescufi** [comment]
@dleach02 @danieldegrasse could you please take a look?

**Review by arnopo — APPROVED** [review]
LGTM

**Review by mmahadevan108 — APPROVED** [review]

**Comment by dleach02** [comment]
@emfend I changed the title for clarity. If you disagree please update the title appropriately.

**Review by danieldegrasse — APPROVED** [review]

**Comment by dbaluta** [comment]
@arnopo @carlescufi How did you tested this with Linux kernel. Looking at the code generic loading FW function still checks for the rsc_table.

https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/remoteproc_core.c#L2477
https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/remoteproc_elf_loader.c#L336

**Comment by arnopo** [comment]
> @arnopo @carlescufi How did you tested this with Linux kernel. Looking at the code generic loading FW function still checks for the rsc_table.
> 
> https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/remoteproc_core.c#L2477 https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/remoteproc_elf_loader.c#L336

I test this sample on the stm32mp1 platform. We load the zephyr Firmware thanks to the Linux remoteproc framework.
The resource table is optional but is recommended to properly initialize the virtio and rpmsg layers.

**Comment by dbaluta** [comment]
Oh, I see. In your stm32mp1 implementation having the rsc_table is optional.

```

static int stm32_rproc_parse_fw(struct rproc *rproc, const struct firmware *fw)
{
»       if (rproc_elf_load_rsc_table(rproc, fw))                                                                                                                                                                   
»       »       dev_warn(&rproc->dev, "no resource table found for this firmware\n");

»       return 0;
}

```

Got it. Thanks. In the generic rproc_elf_load_rsc_table it is not.
